File: 2023/python/day6.py
#!/usr/bin/env python3

from aoc import *


def part1(input: str):
    parse = comb(input, r"\d+", int)
    times = next(parse)
    distance = next(parse)
    answer = 1
    for t, d in zip(times, distance):
        count = 0
        for i in range(t):
            if (i * (t - i)) > d:
                count += 1
        answer *= count
    return answer


def part2(input: str):
    parse = comb(input, r"\d+")
    t = int("".join(next(parse)))
    d = int("".join(next(parse)))
    start = d // t
    while True:
        if (start * (t - start)) > d:
            break
        start += 1
    end = t
    while True:
        if (end * (t - end)) > d:
            break
        end -= 1
    print(f"Distance: {d}")
    start -= 1
    print(f"{start}: {start * (t-start)}")
    start += 1
    print(f"{start}: {start * (t-start)}")
    print(f"{end}: {end * (t-end)}")
    end += 1
    print(f"{end}: {end * (t-end)}")
    print(start, end)
    return end - start


# 25113600
if __name__ == "__main__":
    main()

File: 2023/src/day6.rs
use aoc::*;
use nom::character::complete;

fn ways_to_win((t, d): (i64, i64)) -> i64 {
    let mut x = ((t * t - 4 * d) as f64).sqrt();
    if x.fract() == 0.0 {
        x -= 0.5;
    }
    ((t as f64 + x) / 2f64) as i64 - ((t as f64 - x) / 2f64) as i64
}
#[aoc(day6, part1)]
pub fn part1(input: &str) -> i64 {
    let mut lines = input.lines();
    let times = comb(complete::i64)(lines.next().unwrap());
    let distances = comb(complete::i64)(lines.next().unwrap());
    times
        .into_iter()
        .zip(distances)
        .map(ways_to_win)
        .inspect(|x| debug!("{}", x))
        .product()
}

fn number(input: Option<&str>) -> i64 {
    input.unwrap().chars().fold(0, |acc, c| {
        if let Some(v) = char::to_digit(c, 10) {
            acc * 10 + v as i64
        } else {
            acc
        }
    })
}

#[aoc(day6, part2)]
pub fn part2(input: &str) -> i64 {
    let mut lines = input.lines();
    ways_to_win((number(lines.next()), number(lines.next())))
}
